proof_of_concept/definitions/identifier.py

```diff
@@ -0,0 +1,101 @@
+"""Identifier for various things in the DDM."""
+import re
+from typing import Any, cast, List, Type
+
+
+class Identifier(str):
+    """An identifier.
+
+    An Identifier is a string of any of the following forms:
+
+    1. party:<namespace>:<name>
+    2. party_collection:<namespace>:<name>
+    3. site:<namespace>:<name>
+    4. asset:<namespace>:<name>:<site_namespace>:<site_name>
+    5. asset_collection:<namespace>:<name>
+    6. result:<id_hash>
+
+    This class also accepts a single asterisk an an identifier, as it
+    is used as a wildcard in rules.
+
+    See the Terminology section of the documentation for details.
+    """
+    def __new__(cls: Type['Identifier'], seq: Any) -> 'Identifier':
+        """Create an Identifier.
+
+        Args:
+            seq: Contents, will be converted to a string using str(),
+            then used as the identifier.
+
+        Raises:
+            ValueError: If str(seq) is not a valid identifier.
+        """
+        data = str(seq)
+
+        if data != '*':
+            parts = data.split(':')
```

Comment:
There's the `parts` method for this right?

Reply:
Yes, but we can't call a method on an object before it's been constructed...

Reply:
Never mind...